### PR TITLE
209 add garbage collection between models

### PR DESCRIPTION
**Pull request by david-thrower**
Added proper garbage collection between models.

### CODE REVIEW

**Review by Thunderblok — APPROVED** [review]
Scope touched

.github/workflows/automerge.yml: branch filter updated → 209-add-garbage-collection-between-models

simple_cerebros_random_search.py: added from gc import collect, added two cleanup blocks (tf.keras.backend.clear_session(); collect()), and set neural_network = None post-write

1) Workflow change (automerge.yml)

What changed: branch filter switched from 205-new-branch-for-testing-searialization to 209-add-garbage-collection-between-models.

Notes

✅ Correct to point CI at the new feature branch.

🔧 Suggest making this more future-proof so we don’t have to PR-edit branch names each time:

Use a prefix pattern:

on:
  push:
    branches:
      - main
      - 'feature/**'
      - 'fix/**'
      - '*/add-garbage-collection-between-models'


Or add a pull_request trigger so CI runs on PRs regardless of source branch:

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]


(Optional) If this workflow is meant for automerge, ensure the job actually performs the automerge (e.g., peter-evans/enable-pull-request-automerge or a gh pr merge --auto --squash step) and that permissions: include pull-requests: write. Right now the snippet only shows contents: read.

Decision for David

Keep branch-specific trigger, or convert to pull_request + main only? (I recommend the latter to avoid future branch renames.)

2) Memory hygiene (TensorFlow) changes

What changed (good direction)

Added explicit cleanup at two points in run_moity_permutations:

After model factory creation (before materialize()):
tf.keras.backend.clear_session(); collect()

After result write:
neural_network = None; tf.keras.backend.clear_session(); collect()

Goal: prevent RAM creep across many small runs. 👍

One ordering risk to fix

You call clear_session() after constructing nnf = NeuralNetworkFactory(...). If the factory materializes lazily, we’re mostly safe; but if it captured any Keras global state on instantiation, we risk detaching pieces we just created.

Safer pattern: run the first clear_session() before constructing the next factory (i.e., move those two lines up one step). This guarantees we start each trial from a clean Keras context.

Tear-down completeness

You already null out neural_network. Also ensure any other heavyweight objects created within the trial are dropped before the final clear_session():

Optimizer / loss objects

tf.data datasets/iterators

Callbacks (e.g., TensorBoard); explicitly close any writers

That removes lingering references so GC actually reclaims graphs/buffers.

Acceptance check (what we expect to see)

Run 5–10 serial trials; nvidia-smi shows no monotonic GPU memory climb (peaks bounce within a small band).

If any creep persists on long loops: enable per-trial subprocess execution for this runner (serial in Phase I). That’s the “always works” CUDA-context reset and pairs fine with your current cleanups.

3) Micro-adjusts (optional but low-risk)

Add TF_FORCE_GPU_ALLOW_GROWTH=true at process bootstrap (env) so TF doesn’t pre-grab the whole device heap.

In debug runs, log the current & peak memory after each trial (tf.config.experimental.get_memory_info('GPU:0') if available) so regressions are obvious.

4) Two targeted questions for David

OK to move the first clear_session() to before NeuralNetworkFactory(...)? (Keeps model construction on a clean graph state.)

Do we create any callbacks/datasets/writers inside run_moity_permutations that should also be nulled/closed before the final cleanup? (If yes, I’ll add them in the same style as neural_network = None.)

5) Tiny follow-up (non-blocking)

Add a quick GPU-only integration test that runs 3 micro-trials and asserts peak memory drift < ~200MB. It catches future regressions immediately (skip on CPU CI).
